add check for maximum databases size to migrate

tsharju opened the pull request:
### Proposed changes in this pull request

Adds possibility for checking that the size of the source database is under some limit. This can e.g. used with `--validate` by adding `--dbs-max-total-size <max_size>` to the `pg_migrate` command.

### Type (put an `x` where ever applicable)
- [ ] Bug fix: Link to the issue
- [x] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement
- [ ] Other

### Checklist
Please put an `x` against the checkboxes. Write a small comment explaining if its `N/A` (not applicable)

- [x] All the tests are passing after the introduction of new changes.
- [x] Added tests respective to the part of code I have written.
- [x] Added proper documentation where ever applicable (in code and README.md).

### Optional extra information

Commit:
add check for maximum databases size to migrate

kmichel-aiven commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -852,6 +852,15 @@ def _check_databases(self):
                 else:
                     self.log.info("Database %r already exists in target", dbname)
 
+    def _check_database_size(self, max_size: float):
+        dbs_size = self.source.c(
+            "SELECT SUM(pg_database_size(oid)) FROM pg_catalog.pg_database "
+            "WHERE datname NOT IN ('template1', 'template0')"
+        )[0]["sum"]
```
This should take in account `--filtered-db`, `--skip-table`/`--with-table` and `--no-replicate-extension-tables`.

tsharju replied:
True. Will address. Thanks.

Commit:
make the result of filter_tables optional

We need to know from the result of filter_tables whether there were
any table filtering options given to it or not. It will return None in
a case where no table filtering options were given.

Commit:
remove version 9.5 from supported target versions

kmichel-aiven commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -389,6 +389,18 @@ def mangle_db_name(self, db_name: str) -> str:
+        if only_tables == []:
+            return 0
+        if only_tables is not None:
+            query = "SELECT SUM(pg_total_relation_size(tablename)) AS size FROM UNNEST(%s) AS tablename"
```
For this query, the `dbname` should be passed to `self.c`, because `pg_total_relation_size` takes the table name and looks for this table in the database of the current connection.

It would be good to add specific tests for that function.

tsharju replied:
You're right. I'll add some unit tests there as well.

kmichel-aiven commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -852,6 +865,17 @@ def _check_databases(self):
                 else:
                     self.log.info("Database %r already exists in target", dbname)
 
+    def _check_database_size(self, max_size: float):
```
That function too should have specific tests (it's mocked in the indirect tests of `test_migrate_checks.py`)

tsharju replied:
It's mocked only to assert that it gets called. It should pass the original function as side effect to make sure that it gets called. But I'll add some more tests.

kmichel-aiven commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -389,6 +389,18 @@ def mangle_db_name(self, db_name: str) -> str:
+            query = "SELECT pg_database_size(oid) AS size FROM pg_catalog.pg_database WHERE datname = %s"
+            args = [dbname]
+        result = self.c(query, args=args)
+        return result[0]['size'] or 0
```
nit. `"` instead of `'`

kmichel-aiven commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -735,6 +747,7 @@ def __init__(
         self.log = logging.getLogger(self.__class__.__name__)
         self.source = PGSource(conn_info=source_conn_info, filtered_db=filtered_db, mangle=mangle)
         self.target = PGTarget(conn_info=target_conn_info, filtered_db=filtered_db, mangle=mangle)
+        self.filtered_db = set(filtered_db.split(",")) if filtered_db else None
```
I think this is not used.

tsharju replied:
Thanks. A leftover from the previous version.
